README.md

# ggHoriPlot: horizon plots in ggplot2 <img src='man/images/sticker_ggHoriPlot.png' align="right" height="180" >

<!-- badges: start -->

[![R-CMD-check](https://github.com/rivasiker/ggHoriPlot/actions/workflows/check-standard.yaml/badge.svg)](https://github.com/rivasiker/ggHoriPlot/actions/workflows/check-standard.yaml)
[![CRAN](https://www.r-pkg.org/badges/version/ggHoriPlot)](https://CRAN.R-project.org/package=ggHoriPlot)
[![downloads](https://cranlogs.r-pkg.org/badges/grand-total/ggHoriPlot)](https://CRAN.R-project.org/package=ggHoriPlot)
[![codecov](https://codecov.io/gh/rivasiker/ggHoriPlot/branch/master/graph/badge.svg?token=8V5E63YVM2)](https://app.codecov.io/gh/rivasiker/ggHoriPlot)
<!-- badges: end -->

This package allows building horizon plots in ggplot2. You can learn
more about the package in `vignette("ggHoriPlot")`.

## Installation

You can install `ggHoriPlot` from CRAN via:

``` r
install.packages("ggHoriPlot")
```

You can also install the development version of the package from GitHub
with the following command:

``` r
#install.packages("devtools")
devtools::install_github("rivasiker/ggHoriPlot")
```

## Basic example

Load the libraries:

``` r
library(tidyverse)
library(ggHoriPlot) 
library(ggthemes)
```

Load the dataset and calculate the cutpoints and origin:

``` r
utils::data(climate_CPH)

cutpoints <- climate_CPH  %>% 
  mutate(
    outlier = between(
      AvgTemperature, 
      quantile(AvgTemperature, 0.25, na.rm=T)-
        1.5*IQR(AvgTemperature, na.rm=T),
      quantile(AvgTemperature, 0.75, na.rm=T)+
        1.5*IQR(AvgTemperature, na.rm=T))) %>% 
  filter(outlier)

ori <- sum(range(cutpoints$AvgTemperature))/2
sca <- seq(range(cutpoints$AvgTemperature)[1], 
           range(cutpoints$AvgTemperature)[2], 
           length.out = 7)[-4]

round(ori, 2) # The origin
#> [1] 6.58

round(sca, 2) # The horizon scale cutpoints
#> [1] -12.11  -5.88   0.35  12.81  19.05  25.28
```

Build the horizon plots in `ggplot2` using `geom_horizon()`:

``` r
climate_CPH %>% ggplot() +
  geom_horizon(aes(date_mine, 
                   AvgTemperature,
                   fill = ..Cutpoints..), 
               origin = ori, horizonscale = sca) +
  scale_fill_hcl(palette = 'RdBu', reverse = T) +
  facet_grid(Year~.) +
  theme_few() +
  theme(
    panel.spacing.y=unit(0, "lines"),
    strip.text.y = element_text(size = 7, angle = 0, hjust = 0),
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks.y = element_blank(),
    panel.border = element_blank()
    ) +
  scale_x_date(expand=c(0,0), 
               date_breaks = "1 month", 
               date_labels = "%b") +
  xlab('Date') +
  ggtitle('Average daily temperature in Copenhagen', 
          'from 1995 to 2019')
```

![](man/figures/CPH_climate-1.png)<!-- -->

## Learn more

You can check out the full functionality of `ggHoriPlot` in the
following guides:

-   [Getting
    started](https://rivasiker.github.io/ggHoriPlot/articles/ggHoriPlot.html)
-   [Examples with real
    data](https://rivasiker.github.io/ggHoriPlot/articles/examples.html)


NEWS.md
# ggHoriPlot 1.0.1

Minor bug correction:

* Removed functions from other packages called with `:::`. 
* Added dropped aesthetics argument to `ggproto`.

# ggHoriPlot 1.0.0

* First CRAN release. 


tests/testthat.R
library(testthat)
library(ggHoriPlot)

test_check("ggHoriPlot")


tests/testthat/test-geom_horizon.R

test_that("NAs are not dropped from the data", {
  df <- data.frame(x = 1:5, y = c(1, 2, NA, 4, 5))

  p <- ggplot(df, aes(x, y))+geom_horizon()

  expect_equal(sum(is.na(layer_data(p)$ymax)),
               length(unique(layer_data(p)$group)))
})


test_that("Outliers are removed", {
  df <- data.frame(x = 1:5, y = c(1, 2, 66, 4, 5))

  p <- ggplot(df, aes(x, y))+geom_horizon(rm.outliers = T)

  expect_equal(levels(layer_data(p)$Cutpoints)[1], "[4.33333, 5)")
})


test_that("The origin is valid", {
  df <- data.frame(x = 1:5, y = c(1, 2, 3, 4, 5))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'origin')

  expect_warning(print(p), 'valid origin')
})

test_that("The horizonscale is valid", {
  df <- data.frame(x = 1:5, y = c(1, 2, 3, 4, 5))

  p <- ggplot(df, aes(x, y))+geom_horizon(horizonscale = 'horizonscale')

  expect_warning(print(p), 'valid horizonscale')
})

test_that("The quantiles origin has a valid horizonscale", {
  df <- data.frame(x = 1:5, y = c(1, 2, 3, 4, 5))

  p <- ggplot(df, aes(x, y))+
    geom_horizon(origin = 'quantiles', horizonscale = 2.5)

  expect_warning(print(p),
                 "positive integer when using origin = 'quantiles'")
})


test_that("The midpoint origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'midpoint')

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[7.66667, 9)",  "[6.33333, 7.66667)", "[5, 6.33333)",
                 "[3.66667, 5)",  "[2.33333, 3.66667)", "[1, 2.33333)"  ))
})

test_that("The mean origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'mean')

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[6.56667, 7.9)", "[5.23333, 6.56667)", "[3.9, 5.23333)",
                 "[2.56667, 3.9)", "[1.23333, 2.56667)", "[-0.1, 1.23333)" ))
})

test_that("The median origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'median')

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[5.16667, 6.5)", "[3.83333, 5.16667)", "[2.5, 3.83333)",
                 "[1.16667, 2.5)", "[-0.166667, 1.16667)", "[-1.5, -0.166667)" ))
})

test_that("The cutpoints of min origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'min')

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[7.66667, 9)",  "[6.33333, 7.66667)", "[5, 6.33333)",
                 "[3.66667, 5)",  "[2.33333, 3.66667)", "[1, 2.33333)" ))
})

test_that("The category assignment of min origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'min')

  expect_equal(sum(str_detect(layer_data(p)$group, 'neg')),
               0)
})

test_that("The quantiles origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 'quantiles')

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[7, 9)",   "[5, 7)",   "[2.5, 5)",
                 "[2, 2.5)", "[1.5, 2)", "[1, 1.5)"))
})

test_that("The numeric origin works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(origin = 3)

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[5.66667, 7)", "[4.33333, 5.66667)",  "[3, 4.33333)",
                 "[1.66667, 3)", "[0.333333, 1.66667)", "[-1, 0.333333)"))
})

test_that("The integer horizonscale works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+geom_horizon(horizonscale = 10)

  expect_equal(length(levels(layer_data(p)$Cutpoints)),
               10)
})

test_that("The custom horizonscale works", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+
    geom_horizon(horizonscale = c(1, 3, 6, 8))

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[6, 8)", "[5, 6)", "[3, 5)", "[1, 3)"))
})

test_that("The custom horizonscale works with numeric origin", {
  df <- data.frame(x = 1:10, y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x, y))+
    geom_horizon(horizonscale = c(1, 3, 6, 8),
                 origin = 2)

  expect_equal(levels(layer_data(p)$Cutpoints),
               c("[6, 8)", "[3, 6)", "[2, 3)", "[1, 2)"))
})

test_that("The processed table has the right dimensions", {
  df <- data.frame(x = 1:10,
               xend = (1:10)+0.999,
               y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x = x, y = y))+geom_horizon()

  expect_equal(nrow(layer_data(p)),
               nrow(df)*length(levels(layer_data(p)$Cutpoints)))
})


test_that("The processed table has the right dimensions with xend", {
  df <- data.frame(x = 1:10,
               xend = (1:10)+0.999,
               y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x = x, xend = xend, y = y))+geom_horizon()

  expect_equal(nrow(layer_data(p)),
               nrow(df)*2*length(levels(layer_data(p)$Cutpoints)))
})


test_that("ymin is always 0", {
  df <- data.frame(x = 1:10,
                   y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x = x, y = y))+geom_horizon()

  expect_true(all(layer_data(p)$ymin == 0))
})

test_that("Mirror works", {
  df <- data.frame(x = 1:10,
                   y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x = x, y = y))+geom_horizon(mirror = T)

  expect_true(all(layer_data(p)$ymax == 1))
})


test_that("Reverse works", {
  df <- data.frame(x = 1:10,
                   y = c(1, 3, 5, 8, 1, 2, 6, 9, 2, 2))

  p <- ggplot(df, aes(x = x, y = y))+geom_horizon(reverse = T)

  expect_true(
    all(
      layer_data(p)$ymin[str_detect(layer_data(p)$Categories, 'neg')] == 1))

  expect_true(
    all(
      layer_data(p)$ymin[str_detect(layer_data(p)$Categories, 'pos')] == 0))
})
















tests/testthat/test-stat_horizon.R


test_that("The origin is valid", {
  df <- data.frame(x = 1:5, y = c(1, 2, 3, 4, 5))

  p <- ggplot(df, aes(x, y))+stat_horizon(origin = 'origin')

  expect_warning(print(p), 'valid origin')
})


R/data.R
#' Peaks times for sports and leisure activities
#'
#' A dataset containing the peak time for doing 29 sports
#' and leisure activities.
#'
#' @docType data
#' @format A data frame with 8,092 rows and 3 variables:
#' \describe{
#'   \item{activity}{name of the activity}
#'   \item{time}{time of the day, in hhmm format}
#'   \item{p}{standardized peak}
#' }
#' @source \url{https://github.com/halhen/viz-pub/blob/master/sports-time-of-day/activity.tsv}
#' @source \url{https://eagereyes.org/blog/2017/joy-plots}
"sports_time"


#' Distribution of COVID-19 cases in Asia
#'
#' A dataset containing the geographic distribution of
#' COVID-19 cases in Asia during 2020.
#'
#' @format A data frame with 12,695 rows and 3 variables:
#' \describe{
#'   \item{date_mine}{date of the measurement in yyyy-mm-dd format}
#'   \item{y}{standardized number of cases}
#'   \item{countriesAndTerritories}{countries and territories in Asia}
#' }
#' @source \url{https://www.ecdc.europa.eu/en/publications-data/download-todays-data-geographic-distribution-covid-19-cases-worldwide}
"COVID"

#' Distribution of repeats along the human genome
#'
#' A dataset containing the percentage of simple repeats
#' in 100 kb windows along the human genome (hg38).
#'
#' @format A data frame with 30,885 rows and 4 variables:
#' \describe{
#'   \item{genoName}{chromosome name}
#'   \item{bin}{starting coordinate of window}
#'   \item{bin_2}{end coordinate of window}
#'   \item{p_repeat}{percentage of repeats}
#' }
#' @source \url{https://genome.ucsc.edu/cgi-bin/hgTrackUi?g=rmsk}
"rmsk"

#' Average temperature in major cities of the US
#'
#' A dataset containing the average temperature in Fahrenheit
#' for major cities in the US for the year 2000.
#'
#' @format A data frame with 57,828 rows and 9 variables:
#' \describe{
#'   \item{Region}{geographic region}
#'   \item{Country}{country}
#'   \item{State}{state or territory}
#'   \item{City}{city or town}
#'   \item{Month}{month}
#'   \item{Day}{day}
#'   \item{Year}{year}
#'   \item{AvgTemperature}{average temperature in Fahrenheit}
#'   \item{date_mine}{data in yyyy-mm-dd format}
#' }
#' @source \url{https://www.kaggle.com/sudalairajkumar/daily-temperature-of-major-cities}
#' @source \url{https://benschmidt.org/2014/06/05/optimally-ordering-geographical-entities-in-linear-space/}
"climate_US"


#' Average temperature in Copenhagen
#'
#' A dataset containing the average temperature in degrees Celsius
#' for Copenhagen between the year 1995 and 2019.
#'
#' @format A data frame with 9,132 rows and 9 variables:
#' \describe{
#'   \item{Region}{geographic region}
#'   \item{Country}{country}
#'   \item{State}{state or territory}
#'   \item{City}{city or town}
#'   \item{Month}{month}
#'   \item{Day}{day}
#'   \item{Year}{year}
#'   \item{AvgTemperature}{average temperature in Fahrenheit}
#'   \item{date_mine}{data in yyyy-mm-dd format}
#' }
#' @source \url{https://www.kaggle.com/sudalairajkumar/daily-temperature-of-major-cities}
"climate_CPH"


R/posneg.R

#' Internal function
#'
#' This function calculates the different positive and
#' negative boundaries for each data point
#' based on the origin and the cutpoint vector provided.
#'
#' @importFrom dplyr as_tibble
#'
#' @param y Numeric vector of values to be transformed.
#' @param origin Numeric origin of the horizon plot.
#' @param vec Numeric vector of cutpoints.
#'
#' @return A data frame, which has been transformed with the supplied
#' origin and vector of cuts.
#'
#' @keywords internal


posneg <- function(y, origin, vec) {

  tab <- list()
  pos_lst <- vec[vec >= origin]
  neg_lst <- vec[vec < origin]

  if (length(pos_lst) != 0) {
    test <- origin
    for (i in 1:(length(pos_lst))) {
      min_compare <- pos_lst[i]
      new_col <- unlist(
        lapply(
          y,
          function(x) ifelse(
            x>test,
            abs(min(x,min_compare)-test)/abs(min_compare-test),
            0)))
      tab[[paste0('ypos', i)]] <- new_col
      test <- pos_lst[i]
    }
  }


  if (length(neg_lst) != 0) {
    test <- origin
    for (i in 1:(length(neg_lst))) {
      max_compare <- neg_lst[i]
      new_col <- unlist(
        lapply(
          y,
          function(x) ifelse(
            x<test,
            abs(max(x,max_compare)-test)/abs(max_compare-test),
            0)))
      tab[[paste0('yneg', i)]] <- new_col
      test <- neg_lst[i]

    }
  }

  as_tibble(tab)
}


R/stat_horizon.R

#' @rdname geom_horizon
#' @param geom Used geom, default to 'ribbon'
#'
#' @importFrom ggplot2 layer
#'
#'
#' @export



stat_horizon <- function(mapping = NULL,
                         data = NULL,
                         geom = "ribbon",
                         position = "identity",
                         ...,
                         na.rm = FALSE,
                         show.legend = TRUE,
                         inherit.aes = TRUE,
                         origin = 'midpoint',
                         horizonscale = 6,
                         rm.outliers = FALSE,
                         reverse = FALSE,
                         mirror = FALSE) {
  layer(
    stat = StatHorizon,
    data = data,
    mapping = mapping,
    geom = geom,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      origin = origin,
      horizonscale = horizonscale,
      rm.outliers = rm.outliers,
      reverse = reverse,
      mirror = mirror,
      ...
    )
  )
}


#' @importFrom ggplot2 ggproto Stat
#' @importFrom glue glue
#' @importFrom dplyr tibble
#' @noRd

StatHorizon <- ggproto(
  "StatHorizon",
  Stat,
  compute_group = function(data,
                           scales,
                           origin = 'midpoint',
                           horizonscale = 6,
                           rm.outliers = FALSE,
                           reverse = FALSE,
                           mirror = FALSE) {
    build_horizon_df_2(data, origin, horizonscale, rm.outliers, reverse, mirror)
  },
  compute_layer = function (self, data, params, layout)
  {
    params <- params[intersect(names(params), self$parameters())]
    args <- c(list(data = quote(data), scales = quote(scales)),
              params)
    ggplot2:::dapply(data, "PANEL", function(data) {
      scales <- layout$get_scales(data$PANEL[1])
      tryCatch(
        do.call(self$compute_panel, args),
        error = function(e) {
          warning(
            glue::glue(
              "Computation failed in `{ggplot2:::snake_class(self)}()`:\n{e$message}"
            )
          )
          tibble()
        }
      )
    })
  },
  required_aes = c("x", "y"),
  optional_aes = c("xend"),
  dropped_aes = c("y", "xend"),
  default_aes = aes(fill = ..Categories.., group = ..group..)

)






R/scale_fill_hcl.R


#' General color function
#' @keywords internal
#'
#' @param reverse Logical, indicating whether the palette should be inverted.
#' @param palette A character string, indicating the palette to be used.
#'
#' @return A function, that, given an integer, returns a character vector
#' of colors of the specified palette.
#'
#' @importFrom grDevices hcl.colors

hcl_generic <- function(reverse, palette, ...) {
  function(n) {
    hcl.colors(n, palette=palette, rev=!reverse)
  }
}

#' Create your own discrete scale
#'
#' These functions allow you to specify your own set of mappings from levels in the
#' data to aesthetic values.
#'
#' The functions `scale_colour_manual()`, `scale_fill_manual()`, `scale_size_manual()`,
#' etc. work on the aesthetics specified in the scale name: `colour`, `fill`, `size`,
#' etc. However, the functions `scale_colour_manual()` and `scale_fill_manual()` also
#' have an optional `aesthetics` argument that can be used to define both `colour` and
#' `fill` aesthetic mappings via a single function call (see examples). The function
#' `scale_discrete_manual()` is a generic scale that can work with any aesthetic or set
#' of aesthetics provided via the `aesthetics` argument.
#'
#' @inheritParams ggplot2::scale_fill_discrete
#' @inheritDotParams ggplot2::discrete_scale -expand -position -aesthetics
#' @param palette the name of the palette to generate colors from. A list of all
#' available palettes can be found by running \code{colorspace::hcl_palettes()}
#' (default is "RdYlBu").
#' @param reverse If \code{TRUE}, the order of the colors is reversed (default is \code{FALSE})'
#'
#' @return Scale layer for the fill aesthetic.
#'
#' @importFrom ggplot2 discrete_scale
#' @export

scale_fill_hcl <- function(..., palette = "RdYlBu", reverse = FALSE) {

  pal <- hcl_generic(reverse, palette)
  discrete_scale("fill", 'hcl', palette = pal, ...)

}


R/compute_ori_sca.R


#' Check if a variable is a whole number
#'
#' @param x A number.
#' @param tol Tolerance threshold.
#'
#' @return A logical value indicating whether a number is an integer, given
#' a certain tolerance threshold.
#'
#' @keywords internal

is.wholenumber <-
  function(x, tol = .Machine$double.eps^0.5)  abs(x - round(x)) < tol


#' Compute the origin and scale of the horizon plot
#'
#' @param data A data frame.
#' @inheritParams geom_horizon
#'
#' @return A list with two elements: \code{ori}, which represents the numeric
#' origin of the horizon plot, and \code{vec_cut}, which is a vector with the
#' cutpoints for the horizon plot.
#'
#' @importFrom stats IQR median quantile
#'
#' @keywords internal

compute_ori_sca <- function(data, origin,
                            horizonscale) {

  # Check if horizonscale is valid
  if (!is.numeric(horizonscale)) {
    stop('please, provide a valid horizonscale')
  }

  # If the origin is the median or mean
  if (origin %in% c('median', 'mean', 'midpoint')) {
    # Define origin cutpoint
    ori <- ifelse(origin == 'median',
                  median(data$y[data$outlier], na.rm = TRUE),
                  ifelse(origin == 'mean',
                         mean(data$y[data$outlier], na.rm = TRUE),
                         sum(range(data$y[data$outlier], na.rm = TRUE))/2))
    # If the horizon scale is an integer
    if (length(horizonscale)==1) {
      if (is.wholenumber(horizonscale)) {
        # Save the number of cuts
        ncut <- horizonscale
        # Calculate separation between cutpoints
        sca <- (range(data$y[data$outlier], na.rm = TRUE)[2]-
                  range(data$y[data$outlier], na.rm = TRUE)[1])/horizonscale
        # Calculate cutpoint vectos
        vec_cut <- c()
        # Positive cutpoints
        for (i in 1:(ncut%/%2)) {
          vec_cut <- c(vec_cut, ori+sca*i)
        }
        # Negative cutpoints
        for (i in 1:((ncut+1)%/%2)) {
          vec_cut <- c(vec_cut, ori-sca*i)
        }
        # If the horizon scale is a vector of numbers
      }
    } else {
      # Save the number of cuts
      ncut <- length(horizonscale)
      # Save them as cutpoints directly
      vec_cut <- horizonscale
    }

  } else if (origin == 'quantiles') {
    if ((length(horizonscale)==1) & (is.wholenumber(horizonscale))) {
      ncut <- horizonscale
      ori <- quantile(data$y[data$outlier],
                      (ncut%/%2)/ncut, na.rm = TRUE)
      vec_cut <- c()
      for (i in 0:ncut) {
        if (i != ncut%/%2) {
          vec_cut <- c(vec_cut,
                       quantile(data$y[data$outlier],
                                (i)/ncut, na.rm = TRUE))
        }
      }
    } else {
      stop("the horizonscale should be a positive integer when using origin = 'quantiles'")
    }
    # If the origin is numeric
  } else if (origin == 'min') {
    ncut <- horizonscale
    ori <- min(data$y[data$outlier], na.rm = TRUE)
    sca <- (range(data$y[data$outlier], na.rm = TRUE)[2]-
              range(data$y[data$outlier], na.rm = TRUE)[1])/horizonscale
    vec_cut <- c()
    for (i in 1:ncut) {
      vec_cut <- c(vec_cut, ori+sca*i)
    }
  } else if (is.numeric(origin)) {
    # Save origin cutpoint
    ori <- origin
    if (length(horizonscale)==1) {
      if (is.wholenumber(horizonscale)) {
        ncut <- horizonscale
        sca <- (range(data$y[data$outlier], na.rm = TRUE)[2]-
                  range(data$y[data$outlier], na.rm = TRUE)[1])/horizonscale
        vec_cut <- c()
        # Positive cutpoints
        for (i in 1:(ncut%/%2)) {
          vec_cut <- c(vec_cut, ori+sca*i)
        }
        # Negative cutpoints
        for (i in 1:((ncut+1)%/%2)) {
          vec_cut <- c(vec_cut, ori-sca*i)
        }
      }
    } else {
      ncut <- length(horizonscale)
      vec_cut <- horizonscale
    }
  } else {
    stop('please, provide a valid origin')
  }

  # Sort cutpoints
  vec_cut <- c(sort(vec_cut[vec_cut > ori]),
               rev(sort(vec_cut[vec_cut < ori])))

  list(
    ori = ori,
    vec_cut = vec_cut)

}




R/build_horizon_df_2.R


utils::globalVariables(c("ymin", "ymax", "fill", "y", "outlier", "x", "group",
                         "xend", "value"))

#' Compute group for stat_horizon
#'
#' @param data A data frame.
#' @inheritParams geom_horizon
#'
#' @return A data frame transformed for a horizon plot.
#'
#' @importFrom dplyr mutate between select transmute bind_cols
#' @importFrom tidyr pivot_longer starts_with
#' @importFrom magrittr '%>%'
#' @importFrom stringr str_detect
#' @importFrom stats IQR median quantile setNames
#' @keywords internal

build_horizon_df_2 <- function(data, origin, horizonscale,
                               rm.outliers, reverse, mirror) {

  # Find outliers if necessary
  if (rm.outliers) {
    data <- data %>%
      mutate(
        outlier = between(
          y,
          quantile(y, 0.25, na.rm=TRUE)-1.5*IQR(y, na.rm=TRUE),
          quantile(y, 0.75, na.rm=TRUE)+1.5*IQR(y, na.rm=TRUE)))
  } else {
    data <- data %>%
      mutate(outlier = TRUE)
  }

  # Compute origin and horizonscale
  ori_sca_lst <- compute_ori_sca(
    data, origin,
    horizonscale)
  ori <- ori_sca_lst$ori
  vec_cut <- ori_sca_lst$vec_cut

  # Remove outliers
  data <- select(data, -outlier)

  # Add xend if necessary
  if ('xend' %in% names(data)) {
    data <- data %>%
      pivot_longer(c(x, xend)) %>%
      transmute(x = value, y)
  } else {
    data <- data %>% select(x, y)
  }

  # Modify the data frame by the cutpoints
  data <- data %>%
    bind_cols(posneg(data$y, ori, vec_cut))
  # Tidy up the data frame
  data <- data %>% select(-y) %>% pivot_longer(starts_with('y'))
  colnames(data) <- c("x","fill","value")
  data$ymin<-0
  data$ymax <- data$value
  data <- data %>%
    transmute(x, ymin, ymax, fill, group = fill)
  binnames <- unique(data$group)
  binnames <- c(rev(binnames[str_detect(binnames, 'yneg')]),
                binnames[str_detect(binnames, 'ypos')])
  cutpoint_vec <- rev(sort(c(vec_cut, ori)))
  # Calculate the categories and cutpoints
  data$Categories <- factor(data$group, rev(binnames))
  new_vec <- rep(NA, length(cutpoint_vec)-1)
  for (i in 1:(length(cutpoint_vec)-1)) {
    new_vec[i] <- paste0('[', signif(cutpoint_vec[i+1]), ', ',
                         signif(cutpoint_vec[i]), ')')
  }
  new_vec <- setNames(new_vec, rev(binnames))
  data$Cutpoints <- factor(new_vec[data$group], new_vec)

  # Reverse negative if necessary
  if (reverse) {
    data <- data %>%
      mutate(
        ymin = ifelse(str_detect(group, 'neg'), 1, 0),
        ymax = ifelse(str_detect(group, 'neg'), 1-ymax, ymax)
      )
  }

  # Mirror everything if necessary
  if (mirror) {
    ymin_vec <- data$ymin
    data <- data %>%
      mutate(
        ymin = -ymax+1,
        ymax = ymin_vec+1
      )
  }

  data
}








R/geom_horizon.R

#' Horizon plots
#'
#' This function builds horizon plots in ggplot2. It allows for the
#' customization of the origin and the horizon scale.
#'
#' A horizon plot is a special type of area plot in which the original data
#' is transformed based on an origin and a horizon scale. The data is cut in
#' different intervals, and the further the data is from the origin, the deeper
#' its color usually is. All the intervals above the origin are then stacked on
#' top of one another, keeping the intervals closest to the origin in the bottom
#' and the furthest away ones on top. Likewise, the intervals below the origin
#' are normally given a different color palette and they are stacked in a similar
#' manner in the same area as the intervals above the origin. You can learn more
#' about how horizon plots are built in \code{vignette('ggHoriPlot')} or at
#' \url{https://bernatgel.github.io/karyoploter_tutorial/Tutorial/PlotHorizon/PlotHorizon.html}.
#'
#' @eval ggplot2:::rd_orientation()
#'
#' @section Aesthetics:
#' \code{geom_horizon()} understands the following aesthetics (required aesthetics are in bold):
#' \itemize{
#' \item \strong{\code{x}}
#' \item \strong{\code{y}}
#' \item \code{xend}
#' \item \code{alpha}
#' \item \code{colour}
#' \item \code{fill}
#' }
#'
#' @inheritParams ggplot2::layer
#' @inheritParams ggplot2::geom_bar
#' @param origin Origin of the horizon plot. It can either be a string, namely
#' \code{"midpoint"} (the default), \code{"median"},
#' \code{"mean"}, \code{"min"} or \code{"quantiles"}, or a user-specified number.
#' @param horizonscale Cutpoints of the horizon plot. It can either be an integer
#' specifying the number of ranges (default is \code{6}),
#' or a user-supplied numeric vector with the cutpoints defining the different ranges.
#' @param rm.outliers If \code{TRUE}, all the values below \code{quantile(y, 0.25)-1.5*IQR(y)}
#' and above \code{quantile(y, 0.75)+1.5*IQR(y)} are excluded from the origin and cutpoint
#' calculations (default is \code{FALSE}). @param reverse If \code{TRUE}, the horizon peaks
#' for the values below the origin are reversed (default is \code{FALSE}).
#' @param reverse IF \code{TRUE}, the horizon peaks for all the values below the origin
#' are reversed (default is \code{FALSE}).
#' @param mirror If \code{TRUE}, the horizon peaks for all the values are reversed
#' (default is \code{FALSE}).
#'
#' @return 'ggplot2' layer for building a horizon plot.
#'
#' @examples
#' # Generate data
#' huron <- data.frame(year = 1875:1972, level = as.vector(LakeHuron))
#'
#' # Basic plot with default colors
#' h <- ggplot(huron) + geom_horizon(aes(year, level))
#'
#' # Add color scheme
#' h + theme_void() + scale_fill_hcl()
#'
#' # Add cupoints
#' ggplot(huron) +
#'   geom_horizon(aes(year, level, fill = ..Cutpoints..)) +
#'   theme_void() +
#'   scale_fill_hcl()
#'
#'
#' @importFrom ggplot2 layer ggproto aes GeomRibbon
#'
#' @export


geom_horizon <- function(mapping = NULL, data = NULL,
                         position = "identity",
                         ...,
                         na.rm = FALSE,
                         show.legend = TRUE,
                         origin = 'midpoint',
                         horizonscale = 6,
                         rm.outliers = FALSE,
                         reverse = FALSE,
                         mirror = FALSE,
                         inherit.aes = TRUE) {
  layer(
    data = data,
    mapping = mapping,
    stat = 'horizon',
    geom = GeomHorizon,
    position = position,
    show.legend = show.legend,
    inherit.aes = inherit.aes,
    params = list(
      na.rm = na.rm,
      origin = origin,
      horizonscale = horizonscale,
      rm.outliers = rm.outliers,
      reverse = reverse,
      mirror = mirror,
      ...)
  )
}

#' @rdname geom_horizon
#' @format NULL
#' @usage NULL
#' @importFrom ggplot2 ggproto GeomRibbon aes
#' @export

GeomHorizon <- ggproto("GeomHorizon", GeomRibbon,
                       default_aes = aes(colour = NA, fill = NA,
                                         size = 0.5, linetype = 1,
                                         alpha = NA)
)


